internal/controller/nnf_workflow_controller_helpers.go

```diff
@@ -713,37 +713,86 @@ func (r *NnfWorkflowReconciler) getLustreMgsFromPool(ctx context.Context, pool s
+			return corev1.ObjectReference{}, "", dwsv1alpha2.NewResourceError("unexpected number of allocation sets '%d' for persistent MGS", len(nnfStorage.Spec.AllocationSets)).WithFatal()
+		}
+
+		if len(nnfStorage.Spec.AllocationSets[0].Nodes) != 1 {
+			return corev1.ObjectReference{}, "", dwsv1alpha2.NewResourceError("unexpected number of nodes '%d' in allocation set for persistent MGS", len(nnfStorage.Spec.AllocationSets[0].Nodes)).WithFatal()
+		}
+
+		if nnfStorage.Spec.AllocationSets[0].Nodes[0].Count != 1 {
+			return corev1.ObjectReference{}, "", dwsv1alpha2.NewResourceError("unexpected number of count '%d' in node list for persistent MGS", nnfStorage.Spec.AllocationSets[0].Nodes[0].Count).WithFatal()
+		}
+
+		if len(nnfStorage.Status.MgsAddress) == 0 {
+			return corev1.ObjectReference{}, "", dwsv1alpha2.NewResourceError("no LNid listed for persistent MGS").WithFatal()
+		}
+
+		// If the MGT isn't ready, then don't use it
+		if !nnfStorage.Status.Ready {
+			continue
+		}
+
+		// Find the DWS Storage resource for the Rabbit that the MGT is on
+		storage := &dwsv1alpha2.Storage{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      nnfStorage.Spec.AllocationSets[0].Nodes[0].Name,
+				Namespace: corev1.NamespaceDefault,
+			},
+		}
 
-	if nnfStorage.Spec.FileSystemType != "lustre" {
-		return corev1.ObjectReference{}, "", dwsv1alpha2.NewResourceError("invalid file systems type '%s' for persistent MGS", nnfStorage.Spec.FileSystemType).WithFatal()
+		if err := r.Get(ctx, client.ObjectKeyFromObject(storage), storage); err != nil {
+			return corev1.ObjectReference{}, "", dwsv1alpha2.NewResourceError("could not get storage resource %v", client.ObjectKeyFromObject(storage)).WithError(err)
+		}
+
+		// If the Storage resource says the Rabbit isn't ready, then don't use it
+		if storage.Status.Status != dwsv1alpha2.ReadyStatus {
+			continue
+		}
+
+		// The MGT is healthy from what we can tell, so add it to the map of healthy MGTs
+		healthyMgts[nnfStorage.Status.MgsAddress] = corev1.ObjectReference{
+			Kind:      reflect.TypeOf(dwsv1alpha2.PersistentStorageInstance{}).Name(),
+			Name:      persistentStorage.Name,
+			Namespace: persistentStorage.Namespace,
+		}
 	}
 
-	if len(nnfStorage.Spec.AllocationSets) != 1 {
-		return corev1.ObjectReference{}, "", dwsv1alpha2.NewResourceError("unexpected number of allocation sets '%d' for persistent MGS", len(nnfStorage.Spec.AllocationSets)).WithFatal()
+	// Check to make sure there's at least one MGT we can use
+	if len(healthyMgts) == 0 {
+		return corev1.ObjectReference{}, "", dwsv1alpha2.NewResourceError("").WithUserMessage("no healthy MGSs found for pool: %s", pool).WithMajor()
 	}
 
-	if len(nnfStorage.Status.MgsAddress) == 0 {
-		return corev1.ObjectReference{}, "", dwsv1alpha2.NewResourceError("no LNid listed for persistent MGS").WithFatal()
+	// Choose an MGT at random from the map
+	i := rand.Intn(len(healthyMgts))
+	for mgsAddress, persistentStorageReference := range healthyMgts {
```

Comment:
This sent me off to do some reading, and wow, Go sucks here, doesn't it? It looks like some reasonable help is coming in Go 1.23, but until then you're stuck using "experimental" libraries to make this better.

Reply:
Yeah, I was disappointed in this too.